### ISSUE TITLE
nvm for JXCore (better to include Node also)

### ISSUE THREAD

**Issue by tomByrer** [jxcore/jxcore#137]
It would be handy to A/B different versions of JXCore for various testing (of JXCore itself & how different versions affect an app).
Bonus: Switch between JXCore & Node.

IMHO the GoLang version is a better starting point: https://github.com/coreybutler/nvm

**Comment by listepo** [comment]
+1

**Comment by obastemur** [comment]
@ktrzeciaknubisa could be nice to see if we can help nvm people to bring jxcore support.

**Comment by jasonkarns** [comment]
I'm one of the maintainers of [nodenv](https://github.com/OiNutter/nodenv) (and [node-build](https://github.com/OiNutter/node-build)). I would like to help bring jxcore support to nodenv.

**Comment by ktrzeciaknubisa** [comment]
Hi @jasonkarns . Nice to hear that. We have some unofficial [jxtools](https://github.com/ktrzeciaknubisa/jxtools) package (BTW. which is part of official [jxc](https://www.npmjs.com/package/jxc) helper for jxcore-cordova). The thing is that you can use `jxtools` to download any of precompiled JXcore binaries to the local folder, so that might be useful for you. Example taken from README.md:

```js
var jxtools = require('jxtools');

jxtools.http.downloadJXcore('0.3.0.7', 'sm', function(err, jxfile) {
  console.log(err ? "Err: " + err : "OK: " + jxfile);
});
```

Let me know if you have some other needs.

**Comment by obastemur** [comment]
@jasonkarns what exactly do you need ?

**Comment by jasonkarns** [comment]
I did a manual install (binary) of jxcore to `$(nodenv root)/versions/jxcore` for testing. The first discrepancy I noticed is that the jxcore distribution doesn't put the binary in a `bin/` directory. nodenv works by creating shims for all the executables under `bin/`.

The second thing I did was create a symlink as `bin/node` to `bin/jx`. As a nodenv user (which is designed to *switch* nodes) I would expect that switching to jxcore would make `$ node` fire up jxcore.

Both of these things are easily handled in node-build during the installation process. (Though as a normal user, I would have expected to see `jx` under `bin/` already.)

Lastly, I noticed that jxcore doesn't bundle npm (at least, not as a top level binary). I see that jx has an npm subcommand, but I'm guessing that isn't actually npm but rather, jxcore's own cli interface to the npm registry? Or is it *actually* tunneling commands to an internal npm binary? Would it make sense for nodenv to have a symlink or alias such that `npm` is just `jx npm`?

**Comment by ktrzeciaknubisa** [comment]
@jasonkarns I'm trying to find out what `bin/` folder you're talking about. Do you mean that you would expect this folder inside a zip package taken from downloads page? (you've mentioned that you did a manual install).

> Lastly, I noticed that jxcore doesn't bundle npm

Yes, npm is not bundled within download zip package or jx binary, however at first `jx install` command (same as `jx npm xxx`) the npm (or rather modified version of it) is downloaded and saved into `~/.jx/npm`. JXcore is using it ever since.

**Comment by jasonkarns** [comment]
> Do you mean that you would expect this folder inside a zip package taken from downloads page?

Yes. If you see in the precompiled binary tarballs that nodejs and iojs ship, the binary itself is under `bin/`



> saved into ~/.jx/npm

Agh! jxcore doesn't respect XDG. Is there a reason it is polluting my home directory? Why isn't the npm executable downloaded into jxcore's install directory? Ideally it would be downloaded to `bin/` under jxcore's installation dir. Otherwise you now have jxcore's custom npm binary being shared between multiple jxcore installs, no? (Not to mention completely disregarding a user's XDG configuration: standards.freedesktop.org/basedir-spec/basedir-spec-latest.html)

**Comment by jasonkarns** [comment]
I have a working spike using node-build to install the precompiled binaries, which then automatically creates `./bin/`, symlinks `./bin/node` and `bin/jx` to `./jx`

I'm still concerned about `~/.jx` and the npm issue, but the spike is working.

I'm now working on getting node-build to compile jxcore from source. However, I don't see any published tarballs of the source. I would hope to see a source tarball available (with checksum) at the same downloads page as the binaries. For now, I'm just pulling from github.

A nice to have would be some kind of index.json file similar to https://nodejs.org/dist/index.json.  (I might start opening some feature requests for these things soon)

**Comment by ktrzeciaknubisa** [comment]
As for npm topics you've touched, basically you're right. 

> Why isn't the npm executable downloaded into jxcore's install directory?

Only because currently JXcore on posix platforms have no install directory - so far we've putting the single binary into `/usr/local/bin/`. I know it's not perfect.

> Otherwise you now have jxcore's custom npm binary being shared between multiple jxcore installs, no? 

Yes, although each version of JXcore may need different npm version downloaded. But that is performed transparently by `jx install`. 

> However, I don't see any published tarballs of the source.

Well, tarballs are available under https://github.com/jxcore/jxcore/releases

**Comment by jasonkarns** [comment]
opened #675

**Comment by jasonkarns** [comment]
> Well, tarballs are available under https://github.com/jxcore/jxcore/releases

Indeed. Including links to the relevant tarball from the downloads page would be appreciated, that way a checksum can be included. (and then mirrored)

**Comment by jasonkarns** [comment]
I'm going to open a separate issue to continue this discussion (re: nodenv/node-build) so that I don't hijack the nvm thread any further.

**Comment by jasonkarns** [comment]
node-build now ships with definitions for installing and managing multiple versions of jxcore!

https://github.com/OiNutter/node-build/pull/90
